### DISCUSSION TITLE
Loss suddenly increases

### DISCUSSION THREAD

**Discussion by YukinoshitaLove** [nkicsl/Resfusion#11] [Q&A] [answered]
I came across a problem when training resfusion_restore on the LOLDataset. At the start of the training, the loss decreases as expected. But after 17.6K~18.1K iterations, the loss suddenly explodes to a very large number(~1e9). I tried several times but the same problem occurs, using command: python train_resfusion_restore.py.
![image](https://github.com/user-attachments/assets/de333bd2-b16c-414c-9179-5caa8c5c011b)

**Comment by YukinoshitaLove** [comment]
I printed all the inputs of the denoising module and the predict object of it. They were all fine, only the pred_resnoise went wrong. This indicates that the problem lies with the neural network or optimizer.
```
pred_resnoise = self.denoising_module(x=x_t, time=t, input_cond=X_0_hat)
loss = F.mse_loss(input=pred_resnoise, target=resnoise)
self.log('X_0_hat_max', X_0_hat.max(), on_epoch=True, prog_bar=False, logger=True, sync_dist=True)
self.log('X_0_hat_min', X_0_hat.min(), on_epoch=True, prog_bar=False, logger=True, sync_dist=True)
self.log('resnoise_max', resnoise.max(), on_epoch=True, prog_bar=False, logger=True, sync_dist=True)
self.log('resnoise_min', resnoise.min(), on_epoch=True, prog_bar=False, logger=True, sync_dist=True)
self.log('x_t_max', x_t.max(), on_epoch=True, prog_bar=False, logger=True, sync_dist=True)
self.log('x_t_min', x_t.min(), on_epoch=True, prog_bar=False, logger=True, sync_dist=True)
self.log('t_max', t.float().max(), on_epoch=True, prog_bar=False, logger=True, sync_dist=True)
self.log('t_min', t.float().min(), on_epoch=True, prog_bar=False, logger=True, sync_dist=True)
self.log('pred_resnoise_max', pred_resnoise.max(), on_epoch=True, prog_bar=False, logger=True, sync_dist=True)
self.log('pred_resnoise_min', pred_resnoise.min(), on_epoch=True, prog_bar=False, logger=True, sync_dist=True)
```
![image](https://github.com/user-attachments/assets/dbd85251-d86e-4867-9ae1-cb0d4c3badec)
![image](https://github.com/user-attachments/assets/7252ec13-f044-434c-bdd5-95e369688fbe)
![image](https://github.com/user-attachments/assets/5cf66d8d-cae5-4404-b877-8544501d749e)
![image](https://github.com/user-attachments/assets/a396a8b4-12b3-4723-a663-44d1e2b1a043)
![image](https://github.com/user-attachments/assets/aa5a246a-6b38-4bd8-8a9d-a967096900c3)
![image](https://github.com/user-attachments/assets/51d0d5c1-534b-49e2-819c-ac8ffcbb9867)
![image](https://github.com/user-attachments/assets/85407716-466b-4229-970b-8e3bf7463988)
![image](https://github.com/user-attachments/assets/f5e1da85-6baa-4d1f-be34-a206085076d7)
![image](https://github.com/user-attachments/assets/763f0023-4be1-4aed-ad3a-96165b50dc66)
![image](https://github.com/user-attachments/assets/9b80b18d-88aa-4b6a-92b5-985e4802ded8)

**Reply by YukinoshitaLove** [reply]
I tried gradient clip. There was still a loss explosion, but the training process didn't fail.
```
parser.add_argument('--gradient_clip', default=1, type=float)
trainer = Trainer(
  ...,
  gradient_clip_val=args.gradient_clip,
)
```
![image](https://github.com/user-attachments/assets/68992aeb-6467-4267-a986-a4c016f87348)

**Reply by YukinoshitaLove** [reply, marked as answer]
It seems that the problem lies in the Adam optimizer. `( ᐛ )`
Accordding to:

> https://discuss.pytorch.org/t/loss-suddenly-increases-using-adam-optimizer/11338

> In anycase, the problem with Adam is that it uses moving average in the denominator term. So if the gradients get really small and the whole of denominator will be small. Since the gradients are already small, the denominator results in blowup thus pushing you very far away hence huge loss. You may have a look at [https://openreview.net/forum?id=ryQu7f-RZ 560](https://openreview.net/forum?id=ryQu7f-RZ) . I think there are many recent methods which avert this problem including AMSGrad (in the earlier mentioned paper), Hyper-gradient descent (for Adam) etc. Also look for comments in the openreview forum, there seems to be further discussion on this issue. Unfortunately i do not know of any pytorch implementation of these algorithms.

I replaced the Adam optimizer with AMSGrad (and still used `gradient_clip_val=1`). The problem is solved `ヽ( ຶ▮ ຶ)ﾉ`! It indeed comes from the stabilization issue of the Adam itself. Now the train_loss is normal and we can happily train the resfusion_restore. `(◍•ᴗ•◍)`

```
optimizer = optim.Adam(filter(lambda p: p.requires_grad, self.parameters()), lr=lr,
                               weight_decay=self.hparams.weight_decay, amsgrad=True)
```
![image](https://github.com/user-attachments/assets/64ef968b-ba1b-42a3-a680-5e3818331932)
